kernel/posix/mqueue.c

```diff
@@ -0,0 +1,422 @@
+
+typedef struct mqueue_object {
+	sys_snode_t snode;
+	char *mem_buffer;
+	char *mem_obj;
+	struct k_msgq queue;
+	atomic_t ref_count;
+	char *name;
+} mqueue_object;
+
+typedef struct mqueue_desc {
+	char *mem_desc;
+	mqueue_object *mqueue;
+	u32_t  flags;
+} mqueue_desc;
+
+K_SEM_DEFINE(mq_sem, 1, 1);
+
+/* Initialize the list */
+sys_slist_t mq_list = SYS_SLIST_STATIC_INIT(&mq_list);
+
+s64_t timespec_to_timeoutms(const struct timespec *abstime);
+static mqueue_object *find_in_list(const char *name);
+static s32_t send_message(mqueue_desc *mqd, const char *msg_ptr, size_t msg_len,
+			  s32_t timeout);
+static int receive_message(mqueue_desc *mqd, char *msg_ptr, size_t msg_len,
+			   s32_t timeout);
+static void remove_mq(mqueue_object  *msg_queue);
+
+/**
+ * @brief Open a message queue.
+ *
+ * Number of message queue and descriptor to message queue are limited by
+ * heap size. increase the size through CONFIG_HEAP_MEM_POOL_SIZE.
+ *
+ * See IEEE 1003.1
+ */
+mqd_t mq_open(const char *name, int oflags, ...)
+{
+	va_list va;
+	mode_t mode;
+	mq_attr *attrs = NULL;
+	u32_t msg_size = 0, max_msgs = 0;
+	mqueue_object  *msg_queue;
+	mqueue_desc *msg_queue_desc = NULL, *mqd = (mqueue_desc *)(-1);
+	char *mq_desc_ptr, *mq_obj_ptr, *mq_buf_ptr;
+
+	va_start(va, oflags);
+	if (oflags & O_CREAT) {
+		mode = va_arg(va, mode_t);
+		attrs = va_arg(va, mq_attr*);
+	}
+	va_end(va);
+
+	if (attrs != NULL) {
+		msg_size = attrs->mq_msgsize;
+		max_msgs = attrs->mq_maxmsg;
+	}
+
+	if (name == NULL || ((oflags & O_CREAT) && (msg_size <= 0 ||
+						    max_msgs <= 0))) {
+		errno = EINVAL;
+		return (mqd_t)mqd;
+	}
+
+	if ((strlen(name) + 1)  > CONFIG_MQUEUE_NAMELEN_MAX) {
+		errno = ENAMETOOLONG;
+		return (mqd_t)mqd;
+	}
+
+	/* Check if queue already exists */
+	k_sem_take(&mq_sem, K_FOREVER);
+	msg_queue = find_in_list(name);
+	k_sem_give(&mq_sem);
+
+	if ((msg_queue != NULL) && (oflags & O_CREAT) && (oflags & O_EXCL)) {
+		/* Message queue has alreadey been opened and O_EXCL is set */
+		errno = EEXIST;
+		return (mqd_t)mqd;
+	}
+
+	if (msg_queue == NULL && !(oflags & O_CREAT)) {
+		errno = ENOENT;
+		return (mqd_t)mqd;
+	}
+
+	mq_desc_ptr = k_malloc(sizeof(struct mqueue_desc));
+	if (mq_desc_ptr != NULL) {
+		memset(mq_desc_ptr, 0, sizeof(struct mqueue_desc));
+		msg_queue_desc = (struct mqueue_desc *)mq_desc_ptr;
+		msg_queue_desc->mem_desc = mq_desc_ptr;
+	} else {
+		goto free_mq_desc;
+	}
+
+	/* Allocate mqueue object for new message queue */
+	if (msg_queue == NULL) {
+
+		/* Check for message quantity and size in message queue */
+		if (attrs->mq_msgsize > CONFIG_MSG_SIZE_MAX &&
+		    attrs->mq_maxmsg > CONFIG_MSG_COUNT_MAX) {
+			goto free_mq_desc;
+		}
+
+		mq_obj_ptr = k_malloc(sizeof(mqueue_object));
+		if (mq_obj_ptr != NULL) {
+			memset(mq_obj_ptr, 0, sizeof(mqueue_object));
+			msg_queue = (mqueue_object *)mq_obj_ptr;
+			msg_queue->mem_obj = mq_obj_ptr;
+
+		} else {
+			goto free_mq_object;
+		}
+
+		msg_queue->name = (char *) name;
```

Comment:
You need to duplicate name here.  It's not your storage and POSIX doesn't guarantee that it's a string literal or anything.  It could easily point into the caller's stack and be reused immediately.

Reply:
Hi @andyross , 

Sure I will allocate space for name here and then point "msg_queue->name" to newly allocated space and will copy name into it.
My assumption was app  will provide proper "name" and i can point "msg_queue->name" to name string as @laperie also pointed on same.